### PR TITLE
implement interface for `stop_iter` validation set

### LINKED ISSUE #17: interface for `stop_iter` validation set
`stop_iter` can be used to set how many iterations without an improvement in the objective function occur before training should be halted.

xgboost allows for passing a proportion of the training set to use as an internal validation set via `parameters` (`?details_boost_tree_xgboost`).

When `stop_iter` is supplied (as `early_stopping_rounds`), a `valids` argument is required by lightgbm: `”a list of lgb.Dataset objects, used for validation”`. This seems like a much more cumbersome interface than a proportion—a good portion of bonsai's lightgbm infrastructure is taking care of the formatting for `lgb.Dataset` objects. Allowing for passing a proportion of the training data feels much more straightforward, but a possibly confusing interface given that other `set_engine` arguments are always in the format that the modeling function takes them. As of now, the training set is passed.

### PR DESCRIPTION
**Pull request by simonpcouch**
Closes #17. :)

### CODE REVIEW

File: R/lightgbm.R [hunk 1]
```diff
@@ -185,10 +182,49 @@ process_parallelism <- function(args) {
...
+                         early_stopping_rounds) {
+  n <- nrow(x)
+
+  needs_validation <- !missing_validation || !is.null(early_stopping_rounds)
+
+  if (!needs_validation) {
+    trn_index <- 1:n
+    val_index <- NULL
+  } else {
+    m <- min(floor(n * (1 - validation)) + 1, n - 1)
```
Comment (simonpcouch) [inline R/lightgbm.R:195]: Should we message in the case that a user supplies `stop_iter` (`early_stopping_rounds`) but doesn't supply `validation` (`valids`) to let them know we're going to set aside a small proportion of the training data for validation and point them to the `validation` argument?

Comment (topepo) [reply R/lightgbm.R:195]: no, we don't issue a warning for that.

**Comment by simonpcouch** [comment]
Okay, this should be good to go now, @topepo!

**Comment by simonpcouch** [comment]
Use `drop = FALSE` in subsetting, then 👍  from Max.
